_drafts/assert.md

```diff
@@ -0,0 +1,605 @@
+endeavor, especially if they are used correctly.
+
+### Article Disclaimers
+
+- This post is about runtime asserting, not compile time asserting, such as
+  [`static_assert`](https://en.wikichip.org/wiki/c/assert.h/static_assert).
+- This post is targeting the ever-growing number of **non-safety critical**
+  embedded systems, such as consumer electronics, Class 1 medical devices, home
+  automation devices, etc.
+- The above types of embedded systems are primarily constrained by memory and
+  code space limitations, so the information below takes this into account.
+- The examples below are implemented for the ARM Cortex-M series of MCU's. Many
+  of the concepts can be generalized to other architectures.
+
+## Making the Most of Asserts
+
+To get the most value from asserting, it is important to set up some postmortem
+infrastructure. When an assert takes place (or even a system fault), there is _a
+lot_ of context and data one can capture to help debug these issues without a
+debugger attached to the device. These include but are not limited to:
+
+- the file and line number the assert took place,
+- the registers at the time of assert
+- values or function pointers on the stack, to help reassemble the backtrace
+- the logs that took place before the assert and possibly after the reboot took
+  place
+- for those using MyNewt[^1], ESP-IDF[^2], or have Memfault[^3] integrated, a
+  coredump can be captured at the time of the assert, which will provide a full
+  dump of system memory and registers.
+
+With all of the above, especially coredumps, it becomes advantageous to add as
+many asserts to the system as possible, especially for capturing those 1 in
+10,000 hour bugs. This allows developers to quickly track down and investigate
+bugs affecting users in the field, all without having to reproduce the issue on
+a development board.
+
+## Common Usages of Asserts
+
+Every firmware has places where it is and isn't appropriate to assert. The
+examples below likely apply to most non-safety-critical systems.
+
+### When malloc shouldn't fail
+
+Dyanmic memory allocation is common in embedded systems, especially in today's
+Internet connected consumer electronics. These devices need to be able to
+receive an influx Bluetooth packets, use large HTTP buffers, and possibly load
+assets or fonts into RAM for a user interface. Keeping around statically
+allocated buffers for all of these functions would be irresponsible, especially
+for the devices that measure RAM in kilobytes. In these devices, dynamic memory
+allocation is a must.
+
+There are also places where malloc could fail, and where **it shouldn't**.
+
+For example, let's give an example of dynamically allocated mutexes.
+
+```c
+Mutex *mutex_create(void) {
+  Mutex *mutex = malloc(sizeof(Mutex));
+  ASSERT(mutex);
+  mutex_init(mutex);
+
+  return mutex;
+}
+```
+
+This malloc **should never fail**, because it would mean the system is out of
```

Comment:
Maybe also mention if it did fail and we carried on we'd be using a resource unprotected.

Maybe: also mention a scenario where failure is okay? i.e as a backpressure for some sort of network transport ... could be a bullet point in the "when not to assert" section